lib/currency.js

```diff
@@ -0,0 +1,134 @@
+const { Amount, pkg: hsdPkg } = require('hsd');
+const assert = require('bsert');
+
+const { isChainSupported } = require('./chain');
+
+export const CURRENCY_UNITS = {
+  bitcoin: {
+    currency: 'bitcoin',
+    unit: 'btc',
+    milli: 'mbtc',
+    micro: 'bit',
+    base: 'satoshi',
+  },
+  handshake: {
+    currency: hsdPkg.currency,
+    unit: hsdPkg.unit,
+    milli: `m${hsdPkg.unit}`,
+    micro: `u${hsdPkg.unit}`,
+    base: hsdPkg.base,
+  },
+};
+
+export const CURRENCY_TYPES = {
+  currency: 'currency',
+  unit: 'unit',
+  milli: 'milli',
+  micro: 'micro',
+  base: 'base',
+};
+// bitcoincash units currently same as bitcoin
+CURRENCY_UNITS.bitcoincash = { ...CURRENCY_UNITS.bitcoin };
+
+class Currency extends Amount {
+  /**
+   * Create an object for getting currency units and amounts
+   * @constructor
+   * @param {string} chain - one of supported chain types
+   * see { chain.CHAINS }
+   * @param {(String|Number)?} value
+   * @param {String?} unit
+   * @returns {Currency}
+   */
+  constructor(chain, value, unit) {
+    super(value, unit);
+    const options = new CurrencyOptions(chain);
+    this.chain = options.chain;
+    this.units = options.units;
+  }
+
+  /*
+   * Get actual unit string
+   * @param {string} unit - one of supported unit types
+   * returns {string}
+   */
+  getUnit(unit) {
+    assert(CURRENCY_TYPES[unit], `${unit} not a supported unit type`);
+    return this.units[unit];
+  }
+
+  /**
+   * Get unit string or value.
+   * Overwrites hsd's Amount for more generalized API
+   * @param {String} unit
+   * @param {Boolean?} num
+   * @returns {String|Amount}
+   */
+
+  to(unit, num) {
+    const types = CURRENCY_TYPES;
+    switch (unit) {
+      case types.base:
+        return this.toBase(num);
+      case types.micro:
+      case 'bits':
+        return this.toBits(num);
+      case types.milli:
+        // hsd's `toMilli` has a bug
```

Comment:
Did this turn out to be a bug or is it the fact that there are a different number of do's per hns?

Reply:
After talking with Boyma, it should be `3`. Will be pushing up a fix.